### PR TITLE
Support of data alignment + fixed Scalar typecast of few doubles

### PR DESCRIPTION
**Pull request by cmastalli**
This PR brings data alignment support to Crocoddyl (issue #255). As you might know this is an important feature for boosting the computation on your own hardware.

The main problem was with the data create. We use shared pointers to spawn data in each model. The data creation was highly depended on `make_shared`, however it doesn't honor the data alignment request. To enable this request, we need to use `allocate_shared` with the `Eigen::aligned_allocator`.

Additionally, I have reviewed entirely the code to add proper typecasting of `double` to `Scalar`. @proyan you might want to review these changes carefully.

### CODE REVIEW

File: unittest/test_boxqp.cpp [hunk 1]
```diff
@@ -31,7 +31,7 @@ void test_unconstrained_qp_with_identity_hessian() {
   Eigen::VectorXd gradient = Eigen::VectorXd::Random(nx);
   Eigen::VectorXd lb = -std::numeric_limits<double>::infinity() * Eigen::VectorXd::Ones(nx);
   Eigen::VectorXd ub = std::numeric_limits<double>::infinity() * Eigen::VectorXd::Ones(nx);
-  Eigen::VectorXd xinit(nx);
+  Eigen::VectorXd xinit = Eigen::VectorXd::Random(nx);
```
Comment (wxmerkt) [inline unittest/test_boxqp.cpp:34]: It would have been good to highlight this change in the PR summary... looking over 96 files takes a while and this is quite an important fix to the unit test.

Comment (cmastalli) [reply unittest/test_boxqp.cpp:34]: The main changes are described in the description, I believe.

This modification provides a proper method to initialize the vector. Note that before we're not initializing the values.

Comment (wxmerkt) [reply unittest/test_boxqp.cpp:34]: Yes, I understand the changes well. The summary should serve as a summary though in case someone does not read every PR in full length (it's a lot of files...). This is an important catch-and-fix of a bug of a unit test that is likely to not have tested something useful before, and now works well. So it would be good to include these notes in a summary :-).

File: include/crocoddyl/core/actions/lqr.hxx [hunk 2]
```diff
@@ -44,7 +44,7 @@ void ActionModelLQRTpl<Scalar>::calc(const boost::shared_ptr<ActionDataAbstract>
   } else {
     data->xnext.noalias() = Fx_ * x + Fu_ * u + f0_;
   }
-  data->cost = 0.5 * x.dot(Lxx_ * x) + 0.5 * u.dot(Luu_ * u) + x.dot(Lxu_ * u) + lx_.dot(x) + lu_.dot(u);
+  data->cost = Scalar(0.5) * x.dot(Lxx_ * x) + 0.5 * u.dot(Luu_ * u) + x.dot(Lxu_ * u) + lx_.dot(x) + lu_.dot(u);
```
Comment (proyan) [inline include/crocoddyl/core/actions/lqr.hxx:47]: You missed the scalar casting here.

File: unittest/test_boxqp.cpp (hunk 1 above)
Comment (cmastalli) [reply unittest/test_boxqp.cpp:34]: I didn't think it was significant to point out this specific modification.

I will try to do my best in the next time :)

It would be great if you can try these modifications yourself. Remember that you need to match the same compilation flag (i.e. `-march=native`) in each dependency of Crocoddyl that dependents on Eigen.

File: include/crocoddyl/core/actions/lqr.hxx (hunk 2 above)
Comment (cmastalli) [reply include/crocoddyl/core/actions/lqr.hxx:47]: Nice catch! I will create another PR tomorrow with all your annotations :)

File: include/crocoddyl/core/solvers/box-qp.hpp [hunk 3]
```diff
@@ -29,10 +31,10 @@ struct BoxQPSolution {
   std::vector<size_t> clamped_idx;
 };
 
-// Based on Yuval Tassa's BoxQP
```
Comment (proyan) [inline include/crocoddyl/core/solvers/box-qp.hpp:32]: Why did you remove the doc from here? I found it useful to have this doc to understand what was being developed!

Comment (cmastalli) [reply include/crocoddyl/core/solvers/box-qp.hpp:32]: @wxmerkt implemented using this code. Then I re-wrote it from scratch without checking or reviewing that code. I had to removed earlier, but I forgot it.

Comment (cmastalli) [reply include/crocoddyl/core/solvers/box-qp.hpp:32]: I could add citation to Tassa paper (or better the Projected Newton QP paper that Tassa is citing)

Comment (cmastalli) [reply include/crocoddyl/core/solvers/box-qp.hpp:32]: @wxmerkt what is your opinion about it?

I removed it because this is not a copy-translation of Tassa's code, and it seems to me that we should remove this comment.

Comment (wxmerkt) [reply include/crocoddyl/core/solvers/box-qp.hpp:32]: I compared the implementations a few weeks back and there is little difference: polynomial vs linear line-search (makes no difference in convergence) and LpNorm vs L2Norm for convergence check (no difference in convergence). I think there is a minor change in the Armijo criterion for accepting a step. And of course there's the change from inverse() to LLT. 

We should definitely include a reference (also in general add more comments to code - it's often quite an effort to read Crocoddyl code and follow what's going on) - if you worked off the Projected Newton QP cited in Tassa, that should be included. Otherwise we should leave the Tassa reference.

Comment (cmastalli) [reply include/crocoddyl/core/solvers/box-qp.hpp:32]: I have included a preliminary doxygen documentation of this class. See my PR.

File: include/crocoddyl/core/actions/lqr.hxx (hunk 2 above)
Comment (cmastalli) [reply include/crocoddyl/core/actions/lqr.hxx:47]: Solved this issue and other in diff-lqr in my ongoing PR.
